subsys/usb/host/usbh_shell.c

```diff
@@ -0,0 +1,465 @@
+#include <zephyr/sys/byteorder.h>
+#include <usb/usbh.h>
+
+#include "usbh_internal.h"
+#include <usbh_ch9.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(usbh_shell, CONFIG_USBH_LOG_LEVEL);
+
+const static struct shell *ctx_shell;
+static const struct device *uhc_dev;
+
+static void print_dev_desc(const struct shell *shell,
+			   const struct usb_device_descriptor *const desc)
+{
+	shell_print(shell, "bLength\t\t\t%u", desc->bLength);
+	shell_print(shell, "bDescriptorType\t\t%u", desc->bDescriptorType);
+	shell_print(shell, "bcdUSB\t\t\t%u", desc->bcdUSB);
+	shell_print(shell, "bDeviceClass\t\t%u", desc->bDeviceClass);
+	shell_print(shell, "bDeviceSubClass\t\t%u", desc->bDeviceSubClass);
+	shell_print(shell, "bDeviceProtocol\t\t%u", desc->bDeviceProtocol);
+	shell_print(shell, "bMaxPacketSize0\t\t%u", desc->bMaxPacketSize0);
+	shell_print(shell, "idVendor\t\t%u", desc->idVendor);
+	shell_print(shell, "idProduct\t\t%u", desc->idProduct);
+	shell_print(shell, "bcdDevice\t\t%u", desc->bcdDevice);
+	shell_print(shell, "iManufacturer\t\t%u", desc->iManufacturer);
+	shell_print(shell, "iProduct\t\t%u", desc->iProduct);
+	shell_print(shell, "iSerial\t\t\t%u", desc->iSerialNumber);
+	shell_print(shell, "bNumConfigurations\t\t%u", desc->bNumConfigurations);
+}
+
+static int bazfoo_request(struct usbh_contex *const ctx,
+			  struct uhc_transfer *const xfer,
+			  int err)
+{
+	const struct device *dev = ctx->dev;
+
+	shell_print(ctx_shell, "xfer finished %p, err %d", xfer, err);
+
+	while (!k_fifo_is_empty(&xfer->done)) {
+		struct net_buf *buf;
+
+		buf = net_buf_get(&xfer->done, K_NO_WAIT);
+		if (buf) {
+			if (xfer->ep == USB_CONTROL_EP_IN &&
+			    buf->len == sizeof(struct usb_device_descriptor)) {
+				print_dev_desc(ctx_shell, (void *)buf->data);
+			} else {
+				shell_hexdump(ctx_shell, buf->data, buf->len);
+			}
+
+			uhc_xfer_buf_free(dev, buf);
+		}
+	}
+
+	return uhc_xfer_free(dev, xfer);
+}
+
+static int bazfoo_connected(struct usbh_contex *const uhs_ctx)
+{
+	shell_print(ctx_shell, "USB device connected");
+
+	return 0;
+}
+
+static int bazfoo_removed(struct usbh_contex *const uhs_ctx)
+{
+	shell_print(ctx_shell, "USB device removed");
+
+	return 0;
+}
+
+static int bazfoo_rwup(struct usbh_contex *const uhs_ctx)
+{
+	shell_print(ctx_shell, "Bus remote wakeup event");
+
+	return 0;
+}
+
+static int bazfoo_suspended(struct usbh_contex *const uhs_ctx)
+{
+	shell_print(ctx_shell, "Bus suspended");
+
+	return 0;
+}
+
+static int bazfoo_resumed(struct usbh_contex *const uhs_ctx)
+{
+	shell_print(ctx_shell, "Bus resumed");
+
+	return 0;
+}
+
+USBH_DEFINE_CLASS(bazfoo) = {
+	.request = bazfoo_request,
+	.connected = bazfoo_connected,
+	.removed = bazfoo_removed,
+	.rwup = bazfoo_rwup,
+	.suspended = bazfoo_suspended,
+	.resumed = bazfoo_resumed,
+};
+
+static int cmd_desc_device(const struct shell *sh,
+			   size_t argc, char **argv)
+{
+	uint8_t addr;
+	int err;
+
+	addr = strtol(argv[1], NULL, 16);
```

Comment:
Why base16 for address? Pretty much all documents related to USB describe addresses in decimal.